File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> l=new ArrayList<String>();
        backtrack(n,0,0,"",l);
        return l;

     }
     public void backtrack(int n, int open, int close, String s, List<String>l)
     {
        if (close==n)
        {
            l.add(s);
            return;
        }
        if(open<n)
            backtrack(n,open+1,close, s+"(",l);
        if (close<open)
            backtrack(n,open,close+1, s+")",l);
     }

}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def slove(self,n, opn,close,l,s):
        if close==n:
            l.append(s)
            return
        if opn<n:
            self.slove(n,opn+1,close,l,s+"(")
        if close<opn:
            self.slove(n,opn,close+1,l,s+")")

    def generateParenthesis(self, n: int) -> List[str]:
        l=[]
        self.slove(n,0,0,l,"")
        return l

        